File: solutions/python/cricinfo/scorecard_service.py
from scorecard import Scorecard
from threading import Lock
import itertools

class ScorecardService:
    _instance = None
    _lock = Lock()

    def __init__(self):
        self.scorecards = {}
        self.scorecard_id_counter = itertools.count(1)

    @classmethod
    def get_instance(cls):
        with cls._lock:
            if cls._instance is None:
                cls._instance = cls()
        return cls._instance

    def create_scorecard(self, match):
        scorecard_id = self._generate_scorecard_id(match.get_id())
        scorecard = Scorecard(scorecard_id, match)
        self.scorecards[scorecard_id] = scorecard

    def get_scorecard(self, scorecard_id):
        return self.scorecards.get(scorecard_id)

    def update_score(self, scorecard_id, team_id, score):
        scorecard = self.get_scorecard(scorecard_id)
        if scorecard:
            scorecard.update_score(team_id, score)

    def add_innings(self, scorecard_id, innings):
        scorecard = self.get_scorecard(scorecard_id)
        if scorecard:
            scorecard.add_innings(innings)

    def _generate_scorecard_id(self, match_id):
        scorecard_id = next(self.scorecard_id_counter)
        return f"SC-{match_id}-{scorecard_id:04d}"

File: solutions/csharp/cricinfo/ScorecardService.cs
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Threading;

namespace Cricinfo
{
    public class ScorecardService
    {
        private static ScorecardService instance;
        private readonly ConcurrentDictionary<string, Scorecard> scorecards;
        private int scorecardIdCounter;

        private ScorecardService()
        {
            scorecards = new ConcurrentDictionary<string, Scorecard>();
            scorecardIdCounter = 0;
        }

        public static ScorecardService GetInstance()
        {
            if (instance == null)
            {
                instance = new ScorecardService();
            }
            return instance;
        }

        public void CreateScorecard(Match match)
        {
            string scorecardId = GenerateScorecardId(match.GetId());
            Scorecard scorecard = new Scorecard(scorecardId, match);
            scorecards[scorecardId] = scorecard;
        }

        public Scorecard GetScorecard(string scorecardId)
        {
            scorecards.TryGetValue(scorecardId, out Scorecard scorecard);
            return scorecard;
        }

        public void UpdateScore(string scorecardId, string teamId, int score)
        {
            if (scorecards.TryGetValue(scorecardId, out Scorecard scorecard))
            {
                scorecard.UpdateScore(teamId, score);
            }
        }

        public void AddInnings(string scorecardId, Innings innings)
        {
            if (scorecards.TryGetValue(scorecardId, out Scorecard scorecard))
            {
                scorecard.AddInnings(innings);
            }
        }

        private string GenerateScorecardId(string matchId)
        {
            int scorecardId = Interlocked.Increment(ref scorecardIdCounter);
            return $"SC-{matchId}-{scorecardId:D4}";
        }
    }
}